### PR TITLE
Update sparse-checkout feature to submitted v2

### PR DESCRIPTION
**Pull request by derrickstolee**
This replaces the existing sparse-checkout code with [the version submitted as v2](https://public-inbox.org/git/pull.316.v2.git.gitgitgadget@gmail.com/T/#u).

The biggest change is that it includes a `set` subcommand and NOT an `add` subcommand.

It also requires passing `--stdin` to send the folder list over standard in. By default it will read the arguments.

The stored patterns are slightly different, but this does not affect the interface or interactions by other tools.

This also brings in several commits that are already in the `master` branch of git/git.

The `fixup!` commit fixes an issue with cone patterns and `.gitignore`.

### CODE REVIEW

File: dir.h [hunk 1]
```diff
@@ -12,24 +12,24 @@ struct dir_entry {
...
 };
 
-#define EXC_FLAG_NODIR 1
-#define EXC_FLAG_ENDSWITH 4
-#define EXC_FLAG_MUSTBEDIR 8
-#define EXC_FLAG_NEGATIVE 16
+#define PATTERN_FLAG_NODIR 1
+#define PATTERN_FLAG_ENDSWITH 4
+#define PATTERN_FLAG_MUSTBEDIR 8
+#define PATTERN_FLAG_NEGATIVE 16
```
Comment (wilbaker) [inline dir.h]: Git style question (that I didn't see covered in the code guidelines ... and I realize you're not changing it either but thought I'd ask 😄):

I've seen a lot of spots in the code where `1 <<` is used for flag values.  Do you know if there is a preferred style for flags in general?

File: config.c [hunk 2]
```diff
@@ -1366,22 +1366,16 @@ static int git_default_core_config(const char *var, const char *value, void *cb)
...
-			core_sparse_checkout = result;
+		if (core_virtualfilesystem)
+			core_apply_sparse_checkout = 1;
+		else
+			core_apply_sparse_checkout = git_config_bool(var, value);
+		return 0;
+	}
 
+	if (!strcmp(var, "core.sparsecheckoutcone")) {
+		core_sparse_checkout_cone = git_config_bool(var, value);
```
[the remark this answers is not in the document]
Comment (derrickstolee) [reply config.c]: They should never both be on. VFS for Git uses the virtual filesystem hook, while the other tool uses the sparse-checkout file. These are two incompatible ways to modify the skip-worktree bits.

File: dir.h (hunk 1 above)
Comment (derrickstolee) [reply dir.h]: `(1 << x)` makes it clearer that they are independent flags. this is a really old area of code.

File: builtin/rev-list.c [hunk 3]
```diff
@@ -473,8 +473,10 @@ int cmd_rev_list(int argc, const char **argv, const char *prefix)
 				die(_("object filtering requires --objects"));
 			if (filter_options.choice == LOFC_SPARSE_OID &&
 			    !filter_options.sparse_oid_value)
-				die(_("invalid sparse value '%s'"),
-				    filter_options.filter_spec);
+				die(
```
[the remark this answers is not in the document]
Comment (derrickstolee) [reply builtin/rev-list.c]: This one is not mine, so I'm not sure.

**Pull request merged**